2.0-PROPOSAL/README.md

```diff
@@ -0,0 +1,907 @@
+Completion of that process results in bearer token _activation_. All subsequent
+user interactions with Brigade will present the bearer token as proof of
+authentication. Bearer tokens will expire periodically. (As a point of
+reference, the Azure CLI implements authentication in a similar fashion.)
+
+![Authentication Using OpenID Connect](authentication.png)
+
+N.B.: To prevent the dependency on an identity provider from becoming an
+impedance to evaluation (aka "tire kicking") and local development, as well as
+to facilitate initial setup of Brigade on remote clusters, it is proposed that
+"root" access to Brigade may be selectively enabled / disabled at the time of
+deployment / re-deployment / upgrade.
+
+N.B.: When any _new_ user authenticates with OpenID Connect, a new Brigade
+account is created automatically and federated with the user's identity from the
+trusted provider. Such a user initially has no roles assigned (see
+[authorization](#authorization) section). Roles may be granted by a user manager
+or project admin. It is proposed that users, once registered _cannot_ be deleted
+since nothing prevents a deleted user from re-registering. The benefits of
+_keeping_ but disabling the user account and thereby maintaining the relational
+integrity of the underlying data store seem to outweigh the finality of
+hard-deleting the account. Users who require their Brigade access to be revoked
+can therefore have their accounts _locked_ (instead of deleted) by a user
+manager.
+
+Some non-human users-- such as event gateways-- will utilize service accounts
+that human Brigade administrators may manage directly using the CLI/API. Service
+accounts will authenticate by means of non-expiring but revokable bearer tokens.
+(Note these are _not_ Kubernetes service accounts; but _Brigade_ service
+accounts.)
+
+N.B.: For symmetry with user accounts, service accounts, once created, cannot be
+deleted, but can only be locked if access must be revoked. This also helps to
+maintain the relational integrity of the underlying data store.
+
+The scheduling agent is, necessarily, a _special class of non-human user_ that
+can carry out operations disallowed for _all_ other users (including "root").
+These include updating the observed status of an executing worker or job. Being
+a special class of user, the controller will authenticate to the API using a
+shared secret that is mutually agreed upon at the time of deployment /
+re-deployment / upgrade.
+
+Lastly, every worker (another non-human user) should be uniquely authorized to
+create jobs associated with itself. (i.e. No one _but_ a given worker should
+ever create jobs _for_ that worker.) Given this, every worker will be granted a
+bearer token to facilitate sessionless authenthication to the jobs API only.
+
+To prevent session hijacking by means of root password theft or bearer token
+theft, Brigade API servers in remote clusters should secure API traffic using
```

Comment:
For kicking the tires, will there be a process for provisioning self-signed certs?

Reply:
It's easy enough to make a Helm chart create a self-signed cert and poke it into a secret, so yes. This would be a more sensible default than skipping SSL by default.